### PR TITLE
Fix JSI runtime being unexpectedly deallocated by shared pointer

### PR DESCRIPTION
**Pull request by tsapeta**
# Why

Fixes a crash on launch in bare-expo, introduced in #16777 

# How

Keeping the runtime unique pointer to prevent the runtime from being deallocated when the pointer gets out of scope.

# Test Plan

bare-expo launches without crashes, `JavaScriptRuntimeSpec` tests are passing

### CODE REVIEW

File: packages/expo-modules-core/ios/JSI/EXJavaScriptRuntime.mm [hunk 1]
```diff
@@ -19,35 +19,43 @@
...
 @implementation EXJavaScriptRuntime {
-  std::shared_ptr<jsi::Runtime> _runtime;
+  jsi::Runtime *_runtime;
   std::shared_ptr<react::CallInvoker> _jsCallInvoker;
+
+  /**
+   The unique pointer needs to be held when `EXJavaScriptRuntime` is an owner of the runtime (instead of the bridge).
+   This is necessary in tests and when we want to create independent runtime (when `init` without arguments is used).
+   */
+  std::unique_ptr<jsi::Runtime> _runtimeUniquePtr;
```
Comment (Kudo) [inline packages/expo-modules-core/ios/JSI/EXJavaScriptRuntime.mm:29]: i am not clear for the first time to see the comment and code flow.

maybe we could name it like `_runtimeUniquePtr -> _owningRuntimeForTesting` and `_runtime -> _bridgeRuntime`

other than that, it looks good to me.

Comment (tsapeta) [reply packages/expo-modules-core/ios/JSI/EXJavaScriptRuntime.mm:29]: I'm against naming it "for testing" because it's only how we use it now, but in the future I can imagine we may want to spawn new runtimes like this (multithreading, worklets, workers, ...) in apps too.

The problem is that `make*Runtime` functions return a unique pointer, but we only need to get the raw pointer. We could do `_runtime = makeJSCRuntime().release()`, but then we're responsible for deleting the pointer, so we should also add this 👇 to prevent memory leaks.

```objc
- (void)dealloc
{
  delete _runtime;
}
```

On the other side, we can't do this when the `EXJavaScriptRuntime` is initialized through `initWithRuntime:callInvoker:` because the bridge may still hold a reference to this object.

To sum up, we need `_runtimeUniquePtr` only to keep the smart pointer and prevent it from deallocating the raw pointer.

File: packages/expo-modules-core/ios/JSI/EXJavaScriptRuntime.mm [hunk 2]
```diff
@@ -19,35 +19,43 @@
 using namespace facebook;
 
 @implementation EXJavaScriptRuntime {
-  std::shared_ptr<jsi::Runtime> _runtime;
+  jsi::Runtime *_runtime;
```
Comment (wkozyra95) [inline packages/expo-modules-core/ios/JSI/EXJavaScriptRuntime.mm:22]: Based on our slack conversation I assumed that you need to have raw pointer here, but now I see that it's not. You can keep olde code the only change you would need to make is to:
- `initWithRuntime` needs to take raw pointer(already done in this PR)
- `_runtime = std::shared_ptr<jsi::Runtime>(std::shared_ptr<jsi::Runtime>(), runtime)` in `initWithRuntime`

You can check https://en.cppreference.com/w/cpp/memory/shared_ptr/shared_ptr for an explanation
constructor (8)

**Review by wkozyra95 — APPROVED** [review]

File: packages/expo-modules-core/ios/JSI/EXJavaScriptRuntime.mm (hunk 2 above)
Comment (Kudo) [reply packages/expo-modules-core/ios/JSI/EXJavaScriptRuntime.mm]: the shared_ptr usage is cool. thanks @wkozyra95 for sharing awesome c++ knowledge as always 🤩

**Review by Kudo — APPROVED** [review]
